pkg/operator/controller/controller.go

```diff
@@ -61,7 +63,16 @@ func New(mgr manager.Manager, config Config) (controller.Controller, error) {
 
 // Config holds all the things necessary for the controller to run.
 type Config struct {
-	Client                 client.Client
+	Client client.Client
+	// ServiceMonitor resource may not be registerd by the time 'Client' is created
+	// and since 'Client' will not refresh scheme/mapper fields after creation,
+	// ServiceMonitor GVK may not be recognized.
+	// Enabling 'Client' refresh needs locking as this field can be accessed in parallel
+	// by multiple go routines. Instead we are creating 'ServiceMonitorClient' for operating
```

Comment:
What does "'Client' refresh" refer to? Do you mean writing to this struct's `Client` field? The struct is not shared with other controllers, so you must be talking about synchronizing access within this controller, right? How does using a separate field avoid the the synchronization problem?

Reply:
Yes, Client refresh in this context is to modify the fields in `Client` field. 
I was wrongly assumed that in this controller we are trying to ensure default cluster ingress that uses Client to get and create ingress resource if needed. We are running the controller in single threaded mode so we can modify the Client field and don't need to worry about synchronization issue. I will make this change.
Not related to this, but I'm not sure why we are trying to ensure default cluster ingress by looping every minute. May be this can be part of reconciliation (I thought Dane opened a pr?) or can do watch on default ingress to avoid unnecessary looping for this corner case.

Reply:
> Not related to this, but I'm not sure why we are trying to ensure default cluster ingress by looping every minute. May be this can be part of reconciliation (I thought Dane opened a pr?) or can do watch on default ingress to avoid unnecessary looping for this corner case.

That approach has a bootstrap problem: Reconciliation is triggered by an ingresscontroller-created event, so something else needs to create the default ingresscontroller.  We could perhaps use a synthetic event to the controller, or use a watch on a separate channel source, which the operator could use to trigger the first reconciliation, but using a separate goroutine is simple and good enough for now.